File: packages/contracts/src/endpoints.ts
// WIN-D1 (thinking-windows-port §4.4) — the endpoint-resolution law. A local
// IPC endpoint is: on unix, the LAYOUT path joined under the data root by the
// consumer (node:path.join — byte-for-byte the pre-Windows behavior, which is
// why no unix join lives here); on win32, a named pipe in the flat machine-wide
// \\.\pipe\ namespace, so the name itself must carry what the run directory's
// 0700 boundary carried on unix: a scope. The scope is a hash of the data root,
// which keeps two users' pairs (UA-D10) and two data roots from ever sharing a
// pipe, without leaking the path into a namespace every local process can list.
//
// Both planes must derive the SAME name from the SAME data root, so the
// derivation is deliberately dependency-free and dumb: ASCII-only case fold +
// separator/trailing-slash normalization (Windows paths are case-insensitive
// and arrive via env with whatever casing the spawner had; non-ASCII is left
// untouched because JS and Rust Unicode case folds are not bit-identical and a
// missed fold only costs a redundant scope, never a collision), then FNV-1a
// 64-bit over the UTF-8 bytes. The Rust twin is hand-written in
// field-native/src/endpoints.rs and pinned to this one by
// fixtures/endpoint.vector.json — the layout.vector.json pattern (EL9).

import { APP_ID, SOCKETS } from "./registries";

// APP_ID doubles as the pipe-name brand — one authority, and a change to it is
// a deliberate event the endpoint vector catches (it would orphan every running
// pair's discovery, exactly like moving the data root).
const PIPE_PREFIX = "\\\\.\\pipe\\";

const FNV_OFFSET = 0xcbf29ce484222325n;
const FNV_PRIME = 0x100000001b3n;
const MASK_64 = 0xffffffffffffffffn;

/** FNV-1a 64-bit over the UTF-8 bytes of `input`, as 16 lowercase hex chars.
 * Exported for the fixture test only; callers want `pipeEndpointFor`. */
export function fnv1a64Hex(input: string): string {
  const bytes = new TextEncoder().encode(input);
  let hash = FNV_OFFSET;
  for (const byte of bytes) {
    hash ^= BigInt(byte);
    hash = (hash * FNV_PRIME) & MASK_64;
  }
  return hash.toString(16).padStart(16, "0");
}

/** Canonical form of a data root FOR SCOPE HASHING ONLY (never for the
 * filesystem): `\` → `/`, trailing slashes stripped, ASCII letters folded to
 * lowercase. Two spellings of one Windows directory hash alike; nothing else
 * is promised. */
export function canonicalDataRootForScope(dataRoot: string): string {
  let out = dataRoot.replaceAll("\\", "/");
  while (out.length > 1 && out.endsWith("/")) out = out.slice(0, -1);
  let folded = "";
  for (const ch of out) {
    const code = ch.charCodeAt(0);
    folded += code >= 0x41 && code <= 0x5a ? String.fromCharCode(code + 0x20) : ch;
  }
  return folded;
}

/** The scope tag a data root contributes to its pipe names. */
export function pipeScope(dataRoot: string): string {
  return fnv1a64Hex(canonicalDataRootForScope(dataRoot));
}

/** The win32 endpoint for one of the SOCKETS names under `dataRoot`:
 * `\\.\pipe\vibefield-<scope>-<base>`, where `<base>` is the socket file name
 * minus its `.sock` suffix (`mgmt.sock` → `mgmt`). Unix callers never call
 * this — they join the LAYOUT path as always. */
export function pipeEndpointFor(dataRoot: string, socketFile: string): string {
  if (!socketFile.endsWith(".sock")) {
    throw new Error(`pipeEndpointFor wants a SOCKETS name (*.sock), got: ${socketFile}`);
  }
  const base = socketFile.slice(0, -".sock".length);
  return `${PIPE_PREFIX}${APP_ID}-${pipeScope(dataRoot)}-${base}`;
}

/** True for strings in the named-pipe namespace — the tolerant-reader check
 * for code that logs or validates an endpoint it was handed. */
export function isPipeEndpoint(endpoint: string): boolean {
  return endpoint.startsWith(PIPE_PREFIX);
}

/** Every socket name resolvable through this law — the fixture walks these. */
export const PIPE_SOCKET_NAMES: readonly string[] = Object.values(SOCKETS);

File: packages/field-native/src/endpoints.rs
//! WIN-D1 — the endpoint-resolution law's Rust half. On unix a local IPC
//! endpoint is the LAYOUT path joined under the data root (config::join_layout,
//! unchanged); on win32 it is a named pipe whose name carries a scope hash of
//! the data root, because the flat `\\.\pipe\` namespace has no run-directory
//! boundary to lean on. This module is a HAND-WRITTEN twin of
//! `@vibefield/contracts src/endpoints.ts` — same ASCII-only case fold, same
//! separator/trailing-slash normalization, same FNV-1a 64 — and the two are
//! pinned to one vector (`contracts/fixtures/endpoint.vector.json`, the
//! layout.vector.json pattern) by `tests/endpoint_vector.rs`. A deliberate law
//! change edits both languages and the vector in one commit.

use crate::registries::APP_ID;

pub const PIPE_PREFIX: &str = r"\\.\pipe\";

/// Canonical form of a data root FOR SCOPE HASHING ONLY (never a filesystem
/// path): `\` → `/`, trailing slashes stripped, ASCII letters folded to
/// lowercase. Two spellings of one Windows directory hash alike; non-ASCII is
/// left untouched (JS and Rust Unicode folds are not bit-identical — a missed
/// fold costs a redundant scope, never a collision).
fn canonical_data_root_for_scope(data_root: &str) -> String {
    let slashed = data_root.replace('\\', "/");
    let mut trimmed: &str = &slashed;
    while trimmed.len() > 1 && trimmed.ends_with('/') {
        trimmed = &trimmed[..trimmed.len() - 1];
    }
    trimmed
        .chars()
        .map(|c| {
            if c.is_ascii_uppercase() {
                c.to_ascii_lowercase()
            } else {
                c
            }
        })
        .collect()
}

/// FNV-1a 64-bit over the UTF-8 bytes, 16 lowercase hex chars.
fn fnv1a64_hex(input: &str) -> String {
    let mut hash: u64 = 0xcbf2_9ce4_8422_2325;
    for byte in input.as_bytes() {
        hash ^= u64::from(*byte);
        hash = hash.wrapping_mul(0x0000_0100_0000_01b3);
    }
    format!("{hash:016x}")
}

/// The scope tag a data root contributes to its pipe names.
pub fn pipe_scope(data_root: &str) -> String {
    fnv1a64_hex(&canonical_data_root_for_scope(data_root))
}

/// The win32 endpoint for one of the `registries::sockets` names under
/// `data_root`: `\\.\pipe\vibefield-<scope>-<base>` (`mgmt.sock` → `mgmt`).
/// Callers pass registry constants; anything else is a contract violation.
pub fn pipe_endpoint_for(data_root: &str, socket_file: &str) -> String {
    let base = socket_file.strip_suffix(".sock").unwrap_or_else(|| {
        panic!("pipe_endpoint_for wants a sockets name (*.sock), got {socket_file}")
    });
    format!("{PIPE_PREFIX}{APP_ID}-{}-{base}", pipe_scope(data_root))
}

/// The tolerant-reader check for code handed an endpoint string.
pub fn is_pipe_endpoint(endpoint: &str) -> bool {
    endpoint.starts_with(PIPE_PREFIX)
}
